### PR TITLE
Extend C4 replicator APIs to include Collections & Scopes.

### PR DESCRIPTION
**Pull request by jianminzhao**
Two methods are yet to be defined:
(1) C4Replicator::pendingDocIDs()
(2) C4Replicator::isDocumentPending(slice docID)
I wonder whether we should have concept of DocPath = <scope>.<collection>.<id> so that (1) may return list of pending doc paths, and the argument of (2) may be DocPath.

### CODE REVIEW

File: C/include/c4ReplicatorTypes.h [hunk 1]
```diff
@@ -108,6 +109,7 @@ C4API_BEGIN_DECLS
     /** Information about a document that's been pushed or pulled. */
     typedef struct {
         C4HeapString collectionName;
+        C4HeapString scopeName;
```
Comment (pasin) [inline C/include/c4ReplicatorTypes.h:112]: Should this be C4CollectionSpec for consistency?

File: C/include/c4ReplicatorTypes.h [hunk 2]
```diff
@@ -184,10 +188,32 @@ C4API_BEGIN_DECLS
...
+        // Following options should be encoded into the optionsDictFleed per-collection
+        //#define kC4ReplicatorOptionDocIDs           "docIDs"   ///< Docs to replicate (string[])
+        //#define kC4ReplicatorOptionChannels         "channels" ///< SG channel names (string[])
+        //#define kC4ReplicatorOptionFilter           "filter"   ///< Pull filter name (string)
+        //#define kC4ReplicatorOptionFilterParams     "filterParams"  ///< Pull filter params (Dict[string])
+        //#define kC4ReplicatorOptionSkipDeleted      "skipDeleted" ///< Don't push/pull tombstones (bool)
+        //#define kC4ReplicatorOptionNoIncomingConflicts "noIncomingConflicts" ///< Reject incoming conflicts (bool)
+        //#define kC4ReplicatorCheckpointInterval     "checkpointInterval" ///< How often to checkpoint, in seconds (number)
+        //
+        C4Slice                             optionsDictFleece;
```
Comment (pasin) [inline C/include/c4ReplicatorTypes.h:206]: I don't know if those beside docIDs, Channels, Filter/FilterParams should be configured per collection. Can you check with Jens?

File: C/include/c4ReplicatorTypes.h [hunk 3]
```diff
@@ -184,10 +188,32 @@ C4API_BEGIN_DECLS
 #endif // COUCHBASE_ENTERPRISE
 
 
+    typedef struct C4ReplicationCollection {
+        C4CollectionSpec collection;
+
+        C4ReplicatorMode                    push;              ///< Push mode (from db to remote/other db)
+        C4ReplicatorMode                    pull;              ///< Pull mode (from db to remote/other db).
```
Comment (pasin) [inline C/include/c4ReplicatorTypes.h:195]: I understand that push and pull are per replication not collection.

**Comment by pasin** [comment]
Maybe 

```
C4Replicator::pendingDocIDs(C4CollectionSpec collection)
C4Replicator::isDocumentPending(slice docID, C4CollectionSpec collection)
```

File: C/include/c4ReplicatorTypes.h (hunk 1 above)
Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:112]: C4CollectionSpec is not on heap. I can only be passed around.

Comment (pasin) [reply C/include/c4ReplicatorTypes.h:112]: OK

File: C/include/c4ReplicatorTypes.h (hunk 3 above)
Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:195]: Theoretically, collections can be push/pull independently, but if we spec it that the same replicator shall have the same replicator mode, then we man move it out of collection area.

**Comment by jianminzhao** [comment]
> Maybe
> 
> ```
> C4Replicator::pendingDocIDs(C4CollectionSpec collection) // The API will be per collection on the platform side.
> C4Replicator::isDocumentPending(slice docID, C4CollectionSpec collection)
> ```
This is also my first reaction, but thinking it again, perhaps clients may just want to know what are the pending docs now, without regard to which collections. As collections become popular, merely docID becomes ambiguous. DocPath may be more convenient.
Perhaps we can leave them out for now when Platforms start to write Replicator?

File: C/include/c4ReplicatorTypes.h (hunk 3 above)
Comment (pasin) [reply C/include/c4ReplicatorTypes.h:195]: FYI : The CBL API Spec have it the per replication.

**Comment by pasin** [comment]
FYI : The CBL API will be `getPendingDocIDs(Collection collection)`.

**Review by pasin — APPROVED** [review]
I chatted with @jianminzhao regarding this PR and the main purpose of this PR is to have an API for platforms to start working on. There will be some fine-tuning and discussion about the idea of explicit per-collection option keys and the pending doc IDs API with the Lite-Core team when @jianminzhao is back from vacation. So I approved this PR.

File: C/include/c4ReplicatorTypes.h (hunk 2 above)
Comment (snej) [reply C/include/c4ReplicatorTypes.h:206]: It doesn't seem useful to have the others be per-collection, as far as I can tell.

**Review by snej — CHANGES_REQUESTED** [review]

File: C/include/c4ReplicatorTypes.h (hunk 1 above)
Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:112]: So you did not mean, by using C4HeapString, to allow it to persist? I thought you wanted to keep the slice's buf after the caller deletes it? I can change it to C4CollectionSpec.

File: C/include/c4ReplicatorTypes.h [hunk 4]
```diff
@@ -149,14 +151,16 @@ C4API_BEGIN_DECLS
...
-    typedef bool (*C4ReplicatorValidationFunction)(C4String collectionName,
+    typedef bool (*C4ReplicatorValidationFunction)(C4CollectionSpec collectionSpec,
                                                    C4String docID,
                                                    C4String revID,
                                                    C4RevisionFlags,
                                                    FLDict body,
                                                    void* C4NULLABLE context);
 
 #ifdef COUCHBASE_ENTERPRISE
+    // jzhao - the following two may need collectionSpec, or what documentID should belong to.
```
[the remark this answers is not in the document]
Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:162]: Okay, I simply keep it as is now and remove my comment

File: C/include/c4ReplicatorTypes.h (hunk 3 above)
Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:195]: I can keep it collection specific now. Platforms just need to be little verbose to specify individually.

File: C/include/c4ReplicatorTypes.h (hunk 2 above)
Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:206]: I'll remove others from the above comment.

File: C/include/c4ReplicatorTypes.h (hunk 3 above)
Comment (pasin) [reply C/include/c4ReplicatorTypes.h:195]: Right now that doesn't support by the API. The replicationType is per replicator.

Comment (pasin) [reply C/include/c4ReplicatorTypes.h:195]: LiteCore can make it supported, and we can revise the CBL API. Even we don't support now, support it later will not be breaking change.

Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:195]: Okay, that's fine. I can just simply copy it multiple places.

Comment (jianminzhao) [reply C/include/c4ReplicatorTypes.h:195]: As Platforms convert CBLReplicatorConfig to C4ReplicatorConfig, they may assign the same push/pull mode to all collection area, cannot they?

Comment (pasin) [reply C/include/c4ReplicatorTypes.h:195]: The platform can do that.

**Review by snej — APPROVED** [review]

**Pull request merged**